c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1661,34 +1611,104 @@ std::shared_ptr<void> getIpcDevPtr(std::string handle) {
+      std::sregex_token_iterator end2;
+      std::vector<std::string> kv(it2, end2);
+      if (kv.size() >= 2) {
+        /* Maximum split size in MB.  Limited to large size blocks */
+        if (kv[0].compare("max_split_size_mb") == 0) {
+          size_t val2 = stoi(kv[1]);
+          TORCH_CHECK(
+              val2 > Native::kLargeBuffer / (1024 * 1024),
+              "CachingAllocator option max_split_size_mb too small, must be > ",
+              Native::kLargeBuffer / (1024 * 1024),
+              "");
+          val2 = std::max(val2, Native::kLargeBuffer / (1024 * 1024));
+          val2 = std::min(
+              val2, (std::numeric_limits<size_t>::max() / (1024 * 1024)));
+          m_max_split_size = val2 * 1024 * 1024;
+          used_max_split_size_mb = true;
+        } else if (kv[0].compare("backend") == 0) {
+          TORCH_CHECK(((kv[1].compare("native") == 0) ||
+                       (kv[1].compare("cudaMallocAsync") == 0)),
+                      "Unknown allocator backend, "
+                      "options are native and cudaMallocAsync");
+          m_allocator_backend = kv[1];
+          used_cudaMallocAsync = (kv[1].compare("cudaMallocAsync") == 0);
+          if (used_cudaMallocAsync) {
+#if CUDA_VERSION >= 11040
+            int version;
+            C10_CUDA_CHECK(cudaDriverGetVersion(&version));
+            TORCH_CHECK(version >= 11040,
+                        "backend:cudaMallocAsync requires CUDA runtime "
+                        "11.4 or newer, but cudaDriverGetVersion returned ",
+                        version);
+#else
+            TORCH_CHECK(false,
+                        "backend:cudaMallocAsync requires Pytorch to be built with "
+                        "CUDA 11.4 or newer, but CUDA_VERSION is ",
+                        CUDA_VERSION);
+#endif
+          }
+        } else {
+          TORCH_CHECK(false, "Unrecognized CachingAllocator option: ", kv[0]);
+        }
+      }
+    }
+
+    if (used_max_split_size_mb && used_cudaMallocAsync) {
+      TORCH_WARN("backend:cudaMallocAsync ignores max_split_size_mb");
+    }
   }
-  int device;
-  C10_CUDA_CHECK(cudaGetDevice(&device));
-  void* r = nullptr;
-  caching_allocator.malloc(
-      &r, device, nbytes, cuda::getCurrentCUDAStream(device));
-  return r;
 }
 
-void* raw_alloc_with_stream(size_t nbytes, cudaStream_t stream) {
-  if (nbytes == 0) {
-    return nullptr;
-  }
-  int device;
-  C10_CUDA_CHECK(cudaGetDevice(&device));
-  void* r = nullptr;
-  caching_allocator.malloc(&r, device, nbytes, stream);
-  return r;
+// Public interface
+const char* allocatorBackend() {
+  // Static initializer is thread-safe
+  static const std::string backend = []() {
+                                       if (!parsed) {
+                                         parseArgs();
+                                       }
+                                       return m_allocator_backend;
+                                     }();
```

Comment:
This code seems worse than what we had before. With this, you have to repeat the "if not parsed, parseArgs" logic for every configuration accessor. In the previous version, there was a single `instance()` function that mediated making sure that everything was parsed. Why did you do it this way?

Reply:
I wrote this bit before https://github.com/pytorch/pytorch/pull/71668, so at the time I thought my change was better than the current code :P
At some point during intermittent master merges, I hit intimidating-looking conflicts in CUDACachingAllocator.cpp between my branch and upstream (from https://github.com/pytorch/pytorch/pull/71668 and some other updates) and kept my diffs for this part of the code to be safe. Now that conflicts seem resolved, I agree https://github.com/pytorch/pytorch/pull/71668 is better. I'll rework my code to keep and use https://github.com/pytorch/pytorch/pull/71668's approach.